servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1476,6 +1520,183 @@ public void setState(long fileId, SetStateOptions options)
     }
   }
 
+  /**
+   * Schedules a file for async persistence.
+   *
+   * @param the id of the file for persistence
+   * @throws FileDoesNotExistException when the file does not exist
+   */
+  public synchronized void scheduleAsyncPersistence(long fileId) throws FileDoesNotExistException {
+    // find the worker
+    long workerId = getWorkerStoringFile(fileId);
+
+    // update the state
+    synchronized (mInodeTree) {
+      Inode inode = mInodeTree.getInodeById(fileId);
+      inode.setPersistenceState(FilePersistenceState.SCHEDULED);
+    }
+
+    if (!mWorkerToAsyncPersistFile.containsKey(workerId)) {
+      mWorkerToAsyncPersistFile.put(workerId, Sets.<Long>newHashSet());
+    }
+    mWorkerToAsyncPersistFile.get(workerId).add(fileId);
+  }
+
+  /**
+   * Gets the worker where the given file is stored.
+   *
+   * @param fileId the file id
+   * @return the storing worker
+   * @throws FileDoesNotExistException when the file does not exist on any worker
+   */
+  private long getWorkerStoringFile(long fileId) throws FileDoesNotExistException {
+    List<Long> workers = Lists.newArrayList();
+    try {
+      for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          workers.add(blockLocation.workerId);
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      // should not happen
+      throw new RuntimeException(e);
+    } catch (InvalidPathException e) {
+      // should not happen
+      throw new RuntimeException(e);
+    }
+
+    if (workers.size() == 0) {
+      throw new FileDoesNotExistException("The file " + fileId + " does not exist on any worker");
+    } else if (workers.size() > 1) {
+      LOG.info("the file is stored at more than one worker: " + workers);
+    }
+    return workers.get(0);
```

Comment:
From this code I understand that the `workers` list will contain the workerIds of the workers that have at least one block of the file. Hence my questions:
* What is it that ensures that the worker that gets selected has all of the file blocks (necessary to persist the file)?
* Is returning the first always the best choice? That worker could have just a few block in memory because of a read+recache. Am I correct?